src/vos/vos_aggregate.c

```diff
@@ -1198,7 +1185,8 @@ insert_segments(daos_handle_t ih, struct agg_merge_window *mw,
 		}
 
 		/* Physical entry is in window */
-		if (rect.rc_ex.ex_hi <= mw->mw_ext.ex_hi) {
+		if (rect.rc_ex.ex_hi <= mw->mw_ext.ex_hi ||
+						 phy_ent->pe_retain) {
```

Comment:
I think the physical entry with invalid csum should be kept in queue for next window, otherwise, when it's enqueued again on next window, the sanity check in enqueue_phy_ent() will fail.

Another problem is that it could cause data loss. Let assume a physical entry spans two window, and it's marked as 'pe_retain' in current window because it contains invalid csum in current window (or even it's a good entry, but it's overlapping/adjacent with bad entry, it's also marked as 'pe_retain'), they are silently dropped here, on next window, they will be enqueued again (if it happen to pass the sanity check in enqueue_phy_ent()), and they will be processed in next window (if csum in next window is good, or it was a good entry but marked as pe_retain because it's overlapping with bad one in prior window), and it'll be deleted in next window, so the data in prior window will be lost.

I think there are two options:
1. ideal solution: change 'pe_retain' as 'pe_csum_invalid', and only mark it for the real bad entry (good entry shouldn't be marked), and in fill_one_segment(), adjust segments to exclude the bad portion from bad physical entry.
2. quick hack: if a 'pe_retain' physical entry spans window end, keep it for next window, and make sure the 'pe_retain' flag is kept until it's removed from physical entry queue.

Reply:
I'm not sure I understand how a physical entry that is marked pe_retain because of a csum verification error can be re-enqueued in enqueue_phy_ent().  This function is only called by the iterator call-back.  Also, once a phy_ent is marked for retention (pe_retain set to true), it is never deleted (evt_delete in not called).  I don't see how there can be data loss.

Reply:
This has been fixed.